hack/qemu-metadata-api/server/server.go

```diff
@@ -110,6 +111,43 @@ func (s *Server) listPeers(w http.ResponseWriter, r *http.Request) {
 	log.Infof("Request successful")
 }
 
+// getEndpoint returns the IP address of the first control-plane instance.
+// This allows us to fake a load balancer for QEMU instances.
+func (s *Server) getEndpoint(w http.ResponseWriter, r *http.Request) {
+	log := s.log.With(zap.String("peer", r.RemoteAddr))
+	log.Infof("Serving GET request for /endpoint")
+
+	net, err := s.virt.LookupNetworkByName(s.network)
+	if err != nil {
+		log.With(zap.Error(err)).Errorf("Failed to lookup network")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer net.Free()
+
+	leases, err := net.GetDHCPLeases()
+	if err != nil {
+		log.With(zap.Error(err)).Errorf("Failed to get DHCP leases")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+
+	for _, lease := range leases {
+		if strings.HasPrefix(lease.Hostname, "control-plane") &&
+			strings.HasSuffix(lease.Hostname, "0") {
+			w.Header().Set("Content-Type", "application/json")
+			if err := json.NewEncoder(w).Encode(lease.IPaddr); err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			log.Infof("Request successful")
+			return
+		}
+	}
```

Comment:
Could we return a random control-plane node instead? This is not ideal as we do not perform health checks (as a real loadbalancer would do) but almost gives us high availability for free.

Reply:
I chose to return the first control-plane so we have a consistent Kubernetes endpoint across all nodes. Not sure if/how this will break when we instead return a random control-plane node. I figure it would at least cause problems for the first control-plane node.

Reply:
Yeah. You're probably right. Let's stay consistent!